File: ios/Classes/Events/NotificareEvent.swift
//
//  NotificareEvent.swift
//  notificare
//
//  Created by Helder Pinhal on 03/12/2020.
//

import Foundation
import NotificareKit

enum NotificareEventType: String, CaseIterable {
    case ready
    case deviceRegistered = "device_registered"
}

protocol NotificareEvent {
    var type: NotificareEventType { get }
    var payload: Any? { get }
}

class NotificareEventOnReady: NotificareEvent {
    let type: NotificareEventType
    let payload: Any?

    init(application: NotificareApplication) {
        self.type = .ready
        self.payload = application.toDictionary()
    }
}

class NotificareEventOnDeviceRegistered: NotificareEvent {
    let type: NotificareEventType
    let payload: Any?

    init(device: NotificareDevice) {
        self.type = .deviceRegistered
        self.payload = device.toDictionary()
    }
}

File: notificare-core/android/src/main/kotlin/re/notifica/flutter/events/NotificareEvent.kt
package re.notifica.flutter.events

import re.notifica.flutter.models.toFlutterMap
import re.notifica.models.NotificareApplication
import re.notifica.models.NotificareDevice

internal sealed class NotificareEvent {

    abstract val type: Type
    abstract val payload: Any?

    enum class Type(val id: String) {
        READY(id = "ready"),
        DEVICE_REGISTERED(id = "device_registered"),
    }

    class Ready(
        application: NotificareApplication
    ) : NotificareEvent() {
        override val type = Type.READY
        override val payload = application.toFlutterMap()
    }

    class DeviceRegistered(
        device: NotificareDevice,
    ) : NotificareEvent() {
        override val type = Type.DEVICE_REGISTERED
        override val payload = device.toFlutterMap()
    }
}
